crates/chain-config/src/config/codec.rs

```diff
@@ -0,0 +1,151 @@
+mod decoder;
+mod encoder;
+mod parquet;
+
+pub use decoder::{
+    Decoder,
+    IntoIter,
+};
+pub use encoder::Encoder;
+
+use std::fmt::Debug;
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Group<T> {
+    pub index: usize,
+    pub data: Vec<T>,
+}
+type GroupResult<T> = anyhow::Result<Group<T>>;
+
+#[cfg(test)]
+mod tests {
+    use std::ops::Range;
+
+    use crate::{
+        config::{
+            contract_balance::ContractBalance,
+            contract_state::ContractStateConfig,
+        },
+        CoinConfig,
+        ContractConfig,
+        MessageConfig,
+    };
+
+    use itertools::Itertools;
+
+    use super::*;
+
+    #[test]
+    fn writes_then_reads_written() {
+        let group_size = 100;
+        let num_groups = 10;
+        let starting_group_index = 3;
+        {
+            // Json
+            let temp_dir = tempfile::tempdir().unwrap();
+            let state_encoder = Encoder::json(temp_dir.path());
+
+            let init_decoder = || Decoder::json(temp_dir.path(), group_size).unwrap();
+
+            test_write_read(
+                state_encoder,
+                init_decoder,
+                group_size,
+                starting_group_index..num_groups,
+            );
+        }
+        {
+            // Parquet
+            let temp_dir = tempfile::tempdir().unwrap();
+            let state_encoder = Encoder::parquet(temp_dir.path(), 1).unwrap();
+
+            let init_decoder = || Decoder::parquet(temp_dir.path());
+
+            test_write_read(
+                state_encoder,
+                init_decoder,
+                group_size,
+                starting_group_index..num_groups,
+            );
+        }
+    }
+
+    fn test_write_read(
+        mut encoder: Encoder,
+        init_decoder: impl FnOnce() -> Decoder,
+        group_size: usize,
+        group_range: Range<usize>,
+    ) {
+        let num_groups = group_range.end;
+        let mut rng = rand::thread_rng();
+        macro_rules! write_batches {
```

Comment:
I think this would be more readable as a generic function. Something like.

```rs
fn batches_for_type<T>(rng: &mut impl Rng, count: usize, group_size: usize) -> Vec<T>
   where
        Standard: Distribution<T>,
   {
   ::std::iter::repeat_with(|| rng.gen::<T>())
                    .chunks(group_size)
                    .into_iter()
                    .map(|chunk| chunk.collect_vec())
                    .enumerate()
                    .map(|(index, data)| Group { index, data })
                    .take(count)
                    .collect_vec()
}
```
(IDK if that quite compiles)

and then the test could just be:

```rs
#[test]
fn json_roundtrip_read_write_coins() {
    let mut rng = StdRng::seed_from_u64(42);
    let temp_dir = tempfile::tempdir().unwrap();
    
    // given
    let group_size = 100;
    let num_groups = 10;
    let starting_group_index = 3;
    
    let encoder = Encoder::json(temp_dir.path());
    let decoder = Decoder::json(temp_dir.path(), group_size).unwrap();
    
    let coin_batches = batches_for_type<CoinConfig>(rng, num_groups, group_size);
    
    // when
    encoder.write_coins(coin_batches);
    let read = decoder.coins().unwrap();
    
    // then
    let actual = read.skip(starting_group_index).collect::<Result<Vec<_>, _>>().unwrap();
    let expected = original[starting_group_index..];

    pretty_assertions::assert_eq!(actual, expected);
}
```
Then the test just reads smoothly top-to-bottom. You don't need to look at how everything is working for different tests to understand this one test. Sure, it's less DRY, but much easier to grok quickly.

Reply:
Implemented something very similar to the suggestion. Let me know if you like it :)

Reply:
We've really been pushing for the "given/when/then" format for tests.

 It doesn't have to be exactly my suggestion either. I am curious if you had any specific issues with what I wrote.

Reply:
> I am curious if you had any specific issues with what I wrote.

Nope. When I looked at your suggestion I made the following mental notes:
1. Ok, dislikes macros for readability. DRY what you can without them.
2. Dislikes testing both formats at once. Separating them increases readability.

I didn't even notice the given/when/then suggestion :) 

My only gripe is I dislike two "events" for the "when" phase. So I'll split these tests up into one for each entity and have the encoding be part of the "given" phase -- as in "given something was encoded like so when you decode then... "

> I'm also a fan of tests only testing a single thing
Me as well. The reason why I didn't do it like that is parquet is not human-readable when encoded. Rather than make a test whose assertions/setup are a binary blob, I went with the roundtrip option.

Reply:
>  DRY what you can without them.
DRY is one of my lowest priorities when I code. Especially with tests. It's a nice-to-have, but never at the expense of clarity.

>My only gripe is I dislike two "events" for the "when" phase. So I'll split these tests up into one for each entity and have the encoding be part of the "given" phase -- as in "given something was encoded like so when you decode then... "

Yep. That's what I meant when I said "I'm also a fan of tests only testing a single thing--roundtrip is pushing it already IMO." Roundtrip is kinda an exception to the rule, but I'm never going to complain about splitting it up :).

Reply:
Oh. I see. You kept the roundtrip but only one of the actions is in the `when`.

Is writing not also the subject of these tests?

Reply:
Oh, never mind, I forgot what the tests look like. 

I'll group the actions under `when`. Makes sense since the test is 'roundtrip` after all.